// pip-services-net-node/obj/src/rest/RestService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
let _ = require('lodash');
const pip_services_commons_node_1 = require("pip-services-commons-node");
const pip_services_commons_node_2 = require("pip-services-commons-node");
const pip_services_commons_node_3 = require("pip-services-commons-node");
const pip_services_commons_node_4 = require("pip-services-commons-node");
const pip_services_commons_node_5 = require("pip-services-commons-node");
const HttpResponseSender_1 = require("./HttpResponseSender");
const HttpConnectionResolver_1 = require("./HttpConnectionResolver");
class RestService {
    constructor() {
        this._connectionResolver = new HttpConnectionResolver_1.HttpConnectionResolver();
        this._dependencyResolver = new pip_services_commons_node_2.DependencyResolver(RestService._defaultConfig);
        this._logger = new pip_services_commons_node_3.CompositeLogger();
        this._counters = new pip_services_commons_node_4.CompositeCounters();
    }
    configure(config) {
        config = config.setDefaults(RestService._defaultConfig);
        this._dependencyResolver.configure(config);
        this._connectionResolver.configure(config);
    }
    setReferences(references) {
        this._logger.setReferences(references);
        this._counters.setReferences(references);
        this._dependencyResolver.setReferences(references);
        this._connectionResolver.setReferences(references);
    }
    instrument(correlationId, name) {
        this._logger.trace(correlationId, "Executing %s method", name);
        return this._counters.beginTiming(name + ".exec_time");
    }
    isOpened() {
        return this._server != null;
    }
    open(correlationId, callback) {
        if (this.isOpened()) {
            callback(null);
            return;
        }
        this._connectionResolver.resolve(correlationId, (err, options) => {
            this._uri = options.uri;
            try {
                // Create instance of express application   
                let restify = require('restify');
                this._server = restify.createServer({}); // options);
                // Configure express application
                this._server.use(restify.acceptParser(this._server.acceptable));
                //this._server.use(restify.authorizationParser());
                this._server.use(restify.CORS());
                this._server.use(restify.dateParser());
                this._server.use(restify.queryParser());
                this._server.use(restify.jsonp());
                this._server.use(restify.gzipResponse());
                this._server.use(restify.bodyParser());
                this._server.use(restify.conditionalRequest());
                //this._server.use(restify.requestExpiry());
                // if (options.get("throttle") != null)
                //     this._server.use(restify.throttle(options.get("throttle")));
                this.register();
                this._server.listen(options.port, options.host, (err) => {
                    if (err == null) {
                        // Register the service URI
                        this._connectionResolver.register(correlationId, (err) => {
                            this._logger.debug(correlationId, "Opened REST service at %s", this._uri);
                            if (callback)
                                callback(err);
                        });
                    }
                    else {
                        err = new pip_services_commons_node_5.ConnectionException(correlationId, "CANNOT_CONNECT", "Opening REST service failed")
                            .wrap(err).withDetails("url", this._uri);
                        if (callback)
                            callback(err);
                    }
                });
            }
            catch (ex) {
                this._server = null;
                let err = new pip_services_commons_node_5.ConnectionException(correlationId, "CANNOT_CONNECT", "Opening REST service failed")
                    .wrap(ex).withDetails("url", this._uri);
                if (callback)
                    callback(err);
            }
        });
    }
    close(correlationId, callback) {
        if (this._server != null) {
            // Eat exceptions
            try {
                this._server.close();
                this._logger.debug(correlationId, "Closed REST service at %s", this._uri);
            }
            catch (ex) {
                this._logger.warn(correlationId, "Failed while closing REST service: %s", ex);
            }
            this._server = null;
            this._uri = null;
        }
        callback(null);
    }
    sendResult(req, res) {
        return HttpResponseSender_1.HttpResponseSender.sendResult(req, res);
    }
    sendCreatedResult(req, res) {
        return HttpResponseSender_1.HttpResponseSender.sendCreatedResult(req, res);
    }
    sendDeletedResult(req, res) {
        return HttpResponseSender_1.HttpResponseSender.sendDeletedResult(req, res);
    }
    sendError(req, res, error) {
        HttpResponseSender_1.HttpResponseSender.sendError(req, res, error);
    }
    registerRoute(method, route, schema, action) {
        method = method.toLowerCase();
        if (method == 'delete')
            method = 'del';
        // Hack!!! Wrapping action to preserve prototyping context
        let actionCurl = (req, res) => {
            // Perform validation
            if (schema != null) {
                let params = _.extend({}, req.params, { body: req.body });
                let correlationId = params.correlaton_id;
                let err = schema.validateAndReturnException(correlationId, params, false);
                if (err != null) {
                    this.sendError(req, res, err);
                    return;
                }
            }
            // Todo: perform verification?
            action.call(this, req, res);
        };
        // Wrapping to preserve "this"
        let self = this;
        this._server[method](route, actionCurl);
    }
    register() { }
}
RestService._defaultConfig = pip_services_commons_node_1.ConfigParams.fromTuples("connection.protocol", "http", "connection.host", "0.0.0.0", "connection.port", 3000, "options.request_max_size", 1024 * 1024, "options.connect_timeout", 60000, "options.debug", true);
exports.RestService = RestService;
