// pip-services-net-node/obj/src/rest/RestClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
let _ = require('lodash');
let querystring = require('querystring');
const pip_services_commons_node_1 = require("pip-services-commons-node");
const pip_services_commons_node_2 = require("pip-services-commons-node");
const pip_services_commons_node_3 = require("pip-services-commons-node");
const pip_services_commons_node_4 = require("pip-services-commons-node");
const pip_services_commons_node_5 = require("pip-services-commons-node");
const pip_services_commons_node_6 = require("pip-services-commons-node");
const pip_services_commons_node_7 = require("pip-services-commons-node");
const HttpConnectionResolver_1 = require("./HttpConnectionResolver");
class RestClient {
    constructor() {
        this._connectionResolver = new HttpConnectionResolver_1.HttpConnectionResolver();
        this._logger = new pip_services_commons_node_2.CompositeLogger();
        this._counters = new pip_services_commons_node_3.CompositeCounters();
        this._options = new pip_services_commons_node_1.ConfigParams();
    }
    setReferences(references) {
        this._logger.setReferences(references);
        this._counters.setReferences(references);
        this._connectionResolver.setReferences(references);
    }
    configure(config) {
        config = config.setDefaults(RestClient._defaultConfig);
        this._connectionResolver.configure(config);
        this._options = this._options.override(config.getSection("options"));
    }
    instrument(correlationId, name) {
        this._logger.trace(correlationId, "Executing %s method", name);
        return this._counters.beginTiming(name + ".call_time");
    }
    isOpened() {
        return this._client != null;
    }
    open(correlationId, callback) {
        if (this.isOpened()) {
            if (callback)
                callback(null);
            return;
        }
        this._connectionResolver.resolve(correlationId, (err, options) => {
            if (err) {
                if (callback)
                    callback(err);
                return;
            }
            try {
                this._uri = options.uri;
                let restify = require('restify');
                this._client = restify.createJsonClient({ url: this._uri, version: '*' });
                if (callback)
                    callback(null);
            }
            catch (err) {
                this._client = null;
                let ex = new pip_services_commons_node_6.ConnectionException(correlationId, "CANNOT_CONNECT", "Connection to REST service failed")
                    .wrap(err).withDetails("url", this._uri);
                if (callback)
                    callback(ex);
            }
        });
    }
    close(correlationId, callback) {
        if (this._client != null) {
            // Eat exceptions
            try {
                this._logger.debug(correlationId, "Closed REST service at %s", this._uri);
            }
            catch (ex) {
                this._logger.warn(correlationId, "Failed while closing REST service: %s", ex);
            }
            this._client = null;
            this._uri = null;
        }
        if (callback)
            callback(null);
    }
    addCorrelationId(params, correlationId) {
        // Automatically generate short ids for now
        if (correlationId == null)
            correlationId = pip_services_commons_node_4.IdGenerator.nextShort();
        params = params || {};
        params.correlation_id = correlationId;
        return params;
    }
    addFilterParams(params, filter) {
        params = params || {};
        if (filter) {
            for (let prop in filter) {
                if (filter.hasOwnProperty(prop))
                    params[prop] = filter[prop];
            }
        }
        return params;
    }
    addPagingParams(params, paging) {
        params = params || {};
        if (paging) {
            if (paging.total)
                params.total = paging.total;
            if (paging.skip)
                params.skip = paging.skip;
            if (paging.take)
                params.take = paging.take;
        }
        return params;
    }
    call(method, route, correlationId, params = {}, data, callback) {
        method = method.toLowerCase();
        if (_.isFunction(data)) {
            callback = data;
            data = {};
        }
        params = this.addCorrelationId(params, correlationId);
        if (!_.isEmpty(params))
            route += '?' + querystring.stringify(params);
        let self = this;
        let action = null;
        if (callback) {
            action = (err, req, res, data) => {
                // Handling 204 codes
                if (res.statusCode == 204)
                    callback.call(self, null, null);
                else if (err == null)
                    callback.call(self, null, data);
                else {
                    // Restore application exception
                    if (data != null)
                        err = pip_services_commons_node_5.ApplicationExceptionFactory.create(data).withCause(err);
                    callback.call(self, err, null);
                }
            };
        }
        if (method == 'get')
            this._client.get(route, action);
        else if (method == 'head')
            this._client.head(route, action);
        else if (method == 'post')
            this._client.post(route, data, action);
        else if (method == 'put')
            this._client.put(route, data, action);
        else if (method == 'delete')
            this._client.del(route, action);
        else {
            let error = new pip_services_commons_node_7.UnknownException(correlationId, 'UNSUPPORTED_METHOD', 'Method is not supported by REST client')
                .withDetails('verb', method);
            if (callback)
                callback(error, null);
            else
                throw error;
        }
    }
}
RestClient._defaultConfig = pip_services_commons_node_1.ConfigParams.fromTuples("connection.protocol", "http", "connection.host", "0.0.0.0", "connection.port", 3000, "options.request_max_size", 1024 * 1024, "options.connect_timeout", 60000, "options.debug", true);
exports.RestClient = RestClient;
